File: LeetCode Pratice/words_within_two_edits_of_dictionary.java
import java.util.ArrayList;
import java.util.List;

/**
 * words_within_two_edits_of_dictionary
 */
public class words_within_two_edits_of_dictionary {

  public static void main(String[] args) {
    String[] queries = { "word", "note", "ants", "wood" };
    String[] dictionary = { "wood", "joke", "moat" };
    Innerwords_within_two_edits_of_dictionary sol = new Innerwords_within_two_edits_of_dictionary();
    System.out.println(sol.twoEditWords(queries, dictionary));
  }
}

/**
 * Innerwords_within_two_edits_of_dictionary
 */
class Innerwords_within_two_edits_of_dictionary {

  int difference(String array1, String array2) {
    int maxLength = Math.min(array1.length(), array2.length()); // Compare up to the shorter length
    int differences = 0;

    for (int i = 0; i < maxLength; i++) {
      if (array1.charAt(i) != array2.charAt(i)) {
        differences++;
        if (differences > 2) {
          return differences; // Return early if more than 2 differences are found
        }
      }
    }

    // Account for length differences
    differences += Math.abs(array1.length() - array2.length());

    return differences;
  }

  public List<String> twoEditWords(String[] queries, String[] dictionary) {
    List<String> result = new ArrayList<String>();
    for (int i = 0; i < queries.length; i++) {
      String chs = queries[i];
      for (int j = 0; j < dictionary.length; j++) {
        String inn_chs = dictionary[j];
        int diff_len = difference(chs, inn_chs);
        if (diff_len <= 2) {
          result.add(queries[i]);
          break; // Break inner loop once condition is met
        }
      }
    }
    return result;
  }
}

File: JS/words_within_two_edits_of_dictionary.js
/**
 * @link https://leetcode.com/problems/words-within-two-edits-of-dictionary/description/
 * @param {string[]} queries
 * @param {string[]} dictionary
 * @return {string[]}
 */
function difference(array1, array2) {
  const maxLength = Math.max(array1.length, array2.length);
  let differences = 0;

  for (let i = 0; i < maxLength; i++) {
    if (array1[i] !== array2[i]) {
      differences++;
      if (differences > 2) {
        break;
      }
    }
  }

  return differences;
}
var twoEditWords = function (queries, dictionary) {
  let result = [];
  for (let i = 0; i < queries.length; i++) {
    let chs = queries[i];
    for (let j = 0; j < dictionary.length; j++) {
      let inn_chs = dictionary[j];
      const diff_len = difference(chs, inn_chs);
      if (diff_len <= 2) {
        result.push(queries[i]);
        break;
      }
    }
  }
  return result;
};
const queries = ["word", "note", "ants", "wood"],
  dictionary = ["wood", "joke", "moat"];
console.log(twoEditWords(queries, dictionary));
